File: cpp/hello_world/src/AVLTree.cpp
#include <cstring>
#include <iostream>
using namespace std;

class TreeNode {
public:
  int key;
  TreeNode* left;
  TreeNode* right;
  int height;

  TreeNode(int data) {
    key = data;
    left = NULL;
    right = NULL;
    height = 1;
  }
};

class AVLTree {
public:
  TreeNode* root;

  AVLTree() {
    root = NULL;
  }

  ~AVLTree() {
    destroyNode(root);
  }

  void destroyNode(TreeNode* node) {
    if (node == NULL) { return; }

    destroyNode(node->left);
    destroyNode(node->right);

    delete node;
  }

  int height(TreeNode* node) {
    if (node == NULL) { return 0; }
    return node->height;
  }

  int max(int a, int b) {
    return a > b ? a : b;
  }

  int getBalance(TreeNode* node) {
    return height(node->left) - height(node->right);
  };

  TreeNode* insert(int key) {
    TreeNode* node = insert(root, key);

    if (root == NULL){
      root = node;
    }

    return node;
  }

  void updateHeight(TreeNode* node) {
    node->height = 1 + max(height(node->left),
                           height(node->right));
  }

  TreeNode* rightRotation(TreeNode* node) {
    cout << "rightRotation ";
    TreeNode* newRoot = node->left;
    TreeNode* child = newRoot->right;

    newRoot->right = node;
    node->left = child;

    updateHeight(node);
    updateHeight(newRoot);

    return newRoot;
  }

  TreeNode* leftRotation(TreeNode* node) {
    cout << "leftRotation " << node->key << endl;
    TreeNode* newRoot = node->right;
    TreeNode* child = newRoot->left;
    newRoot->left = node;
    node->right = child;

    updateHeight(node);
    updateHeight(newRoot);

    return newRoot;
  }

  TreeNode* rebalance(TreeNode* node, int key) {
    int balance = getBalance(node);

    // 4 cases
    if (balance > 1) {
      // LEFT LEFT
      if (key < node->left->key) {
        // in left subtree of left node
        return rightRotation(node);
      }

      // LEFT RIGHT
      if (key > node->left->key) {
        // in right subtree of left node
        node->left = leftRotation(node->left);
        return rightRotation(node);
      }
    }

    if (balance < -1) {
      // RIGHT RIGHT
      if (key > node->right->key) {
        return leftRotation(node);
      }

      // RIGHT LEFT
      if (key > node->left->key) {
        node->right = rightRotation(node->right);
        return leftRotation(node);
      }
    }

    return node;
  }

  TreeNode* insert(TreeNode* node, int key) {
    if (node == NULL) {
      return new TreeNode(key);
    }

    if (key < node->key) {
      node->left = insert(node->left, key);
    } else if (key > node->key) {
      node->right = insert(node->right, key);
    } else {
      // Cannot have duplicates
      return node;
    }

    /* 2. Update height of this ancestor node */
    updateHeight(node);
    return rebalance(node, key);
  }

  TreeNode* successor(TreeNode* root) {
    if (root == NULL) { return NULL; }
    root = root->right;
    while (root->left != NULL) {
      root = root->left;
    }
    return root;
  }

  TreeNode* deleteNode(TreeNode* root, int key) {
    if (root == NULL) { return NULL; }
    if (key < root->key) {
      root->left = deleteNode(root->left, key);
    } else if (key > root->key) {
      root->right = deleteNode(root->right, key);
    } else {
      if (root->left == NULL && root->right == NULL) {
        delete root;
        return NULL;
      } else if(root->left == NULL || root->right == NULL) {
        TreeNode* child = root->right == NULL ? root->left : root->right;
        delete root;
        return child;
      } else {
        TreeNode* next = successor(root);
        root->key = next->key;
        root->right = deleteNode(root->right, next->key);
      }
    }

    updateHeight(root);
    return rebalance(root, key);
  }

  TreeNode* deleteNode(int key) {
    return deleteNode(root, key);
  }

  void preOrderPrint() {
    preOrderPrint(root);
  }

  void inOrderPrint() {
    inOrderPrint(root);
  }

  void printNode(TreeNode* node) {
    cout << node->key << "(" << getBalance(node) << ") ";
  }

  void preOrderPrint(TreeNode* node) {
    if (node == NULL) {
      cout << "X ";
      return;
    }

    if (node->left == NULL && node->right == NULL) {
      printNode(node);
      cout << " | ";
      return;
    }

    printNode(node);
    preOrderPrint(node->left);
    preOrderPrint(node->right);
  }

  void inOrderPrint(TreeNode* node) {
    if (node == NULL) {
      cout << "X ";
      return;
    }

    if (node->left == NULL && node->right == NULL) {
      printNode(node);
      cout << " | ";
      return;
    }

    inOrderPrint(node->left);
    printNode(node);
    inOrderPrint(node->right);
  }
};

File: javascript/2019/avlTree.js
// AVL Tree
//jshint esnext: true

class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
    this.height = 1;
  }
  
  getHeight(node) {
    if (!node) { return 0; }
    return node.height;
  }
  updateHeight() {
    this.height = 1 + Math.max(this.getHeight(this.left), this.getHeight(this.right));
  }
  balance() {
    return this.getHeight(this.left) - this.getHeight(this.right);
  }
}

class AVLTree {
  constructor() {
    this.root = null;
  }
  insert(value) {
    this.root = this.doInsert(this.root, value);
  }
  doInsert(root, value) {
    if (!root) {
      return new Node(value);
    }
    if (root.value === value) { return root; } // no duplicates
    if (value < root.value) {
      root.left = this.doInsert(root.left, value);
    }
    if (value > root.value) {
      root.right = this.doInsert(root.right, value);
    }
    root.updateHeight();
    return this.balance(root, value);
  }
  rightRotation(node) {
    let newRoot = node.left;
    let oldChild = newRoot.right;
    node.left = oldChild;
    newRoot.right = node;
    node.updateHeight();
    newRoot.updateHeight();
    return newRoot;
  }
  leftRotation(node) {
    let newRoot = node.right;
    let oldChild = newRoot.left;
    node.right = oldChild;
    newRoot.left = node;
    node.updateHeight();
    newRoot.updateHeight();
    return newRoot;
  }
  balance(root, value) {
    let nodeBalance = root.balance();
    console.log(`balance ${nodeBalance}`);

    if (nodeBalance > 1) {
      if (value < root.left.value) {
        console.log(`LEFT LEFT ${value} ${root.value}`);
        // LEFT LEFT
        return this.rightRotation(root);
      } else if (value > root.left.value) {
        console.log(`LEFT RIGHT ${value} ${root.value}`);
        // LEFT RIGHT
        root.left = this.leftRotation(root.left);
        return this.rightRotation(root);
      }
    } else if (nodeBalance < -1) {
      if (value > root.right.value) {
        console.log(`RIGHT RIGHT ${value} ${root.value}`);
        // RIGHT RIGHT
        return this.leftRotation(root);
      } else if (value < root.right.value) {
        console.log(`RIGHT LEFT ${value} ${root.value}`);
        // RIGHT LEFT
        root.right = this.rightRotation(root.right);
        return this.leftRotation(root);
      }
    }
    
    return root;
  }
  delete(value) {
    this.doDelete(this.root, value);
  }
  
  inOrderSuccessor(node) {
    // return left-most child in right subtree
    let current = node.right;
    while(current.left) {
      current = current.left;
    }
    return current;
  }
  
  doDelete(root, value) {
    if (!root) { return; }
    if (value < root.value) {
      root.left = this.doDelete(root.left, value);
    } else if (value > root.value) {
      root.right = this.doDelete(root.right, value);
    } else {
      if (!root.left && !root.right) { // leaf;
        return null;
      } else if (!root.left) {
        return root.right;
      } else if (!root.right) {
        return root.left;
      } else { // two children
        let next = this.inOrderSuccessor(root);
        root.value = next.value;
        root.right = this.doDelete(root.right, next.value);
      }
    }
    
    root.updateHeight();
    return this.balance(root, value);
  }

  preOrder(root, result = []) {
    if (!root) { return; }
    result.push(root.value);
    if (!root.left && !root.right) { result.push("X"); return; }
    this.preOrder(root.left, result);
    this.preOrder(root.right, result);
    return result;
  }
  inOrder(root, result = []) {
    if (!root) { return; }
    this.inOrder(root.left, result);
    result.push(root.value);
    this.inOrder(root.right, result);
    return result;
  }
}

let t = new AVLTree();
[10, 5, 2,4,3,1].forEach(k => t.insert(k));
console.debug(t.preOrder(t.root));
t.delete(5);
console.debug(t.preOrder(t.root));
